File: resources/engine/ir.py
# -*- coding: utf-8 -*-
"""
Структурный разбор байткода метода в список объектов Instruction.
В отличие от disassembler.py (который сразу печатает текст), здесь каждая
инструкция - это объект с полями, пригодными для дальнейшего анализа
(построение CFG, символическая интерпретация стека).
"""
import struct
from opcodes import OPCODES, NEWARRAY_TYPES


def _u2(b, p):
    return struct.unpack_from(">H", b, p)[0]

def _s1(b, p):
    return struct.unpack_from(">b", b, p)[0]

def _s2(b, p):
    return struct.unpack_from(">h", b, p)[0]

def _s4(b, p):
    return struct.unpack_from(">i", b, p)[0]


class Instruction:
    __slots__ = ("pc", "next_pc", "opcode", "mnemonic", "kind",
                 "ival", "cp_index", "target", "targets", "iinc_idx",
                 "iinc_const", "atype", "count", "dims", "raw_len")

    def __init__(self, pc, opcode, mnemonic, kind):
        self.pc = pc
        self.opcode = opcode
        self.mnemonic = mnemonic
        self.kind = kind
        self.next_pc = None
        self.ival = None          # generic integer operand (byte/short/local index)
        self.cp_index = None      # constant-pool operand
        self.target = None        # absolute pc, for simple branches
        self.targets = None       # dict for switch: {value_or_None(default): abs_pc}
        self.iinc_idx = None
        self.iinc_const = None
        self.atype = None
        self.count = None         # invokeinterface arg count
        self.dims = None          # multianewarray dims

    def __repr__(self):
        return f"<{self.pc}:{self.mnemonic}>"

    @property
    def is_branch(self):
        return self.kind in ("branch2", "branch4")

    @property
    def is_switch(self):
        return self.kind in ("tableswitch", "lookupswitch")

    @property
    def is_return(self):
        return self.mnemonic in ("return", "ireturn", "lreturn", "freturn",
                                  "dreturn", "areturn")

    @property
    def is_unconditional(self):
        return self.mnemonic in ("goto", "goto_w") or self.is_return or \
               self.mnemonic in ("athrow",)

    @property
    def is_conditional(self):
        return self.is_branch and self.mnemonic not in ("goto", "goto_w")


def decode_method(code: bytes):
    """Возвращает dict pc -> Instruction, разбирая весь метод."""
    instrs = {}
    order = []
    n = len(code)
    pc = 0
    while pc < n:
        start = pc
        opcode = code[pc]
        entry = OPCODES.get(opcode)
        if entry is None:
            ins = Instruction(start, opcode, f"<unknown 0x{opcode:02x}>", "none")
            ins.next_pc = start + 1
            instrs[start] = ins
            order.append(start)
            pc += 1
            continue
        mnemonic, kind = entry
        pc += 1
        ins = Instruction(start, opcode, mnemonic, kind)

        if kind == "none":
            pass
        elif kind == "byte":
            ins.ival = _s1(code, pc); pc += 1
        elif kind == "short":
            ins.ival = _s2(code, pc); pc += 2
        elif kind == "ubyte_cp":
            ins.cp_index = code[pc]; pc += 1
        elif kind == "ushort_cp":
            ins.cp_index = _u2(code, pc); pc += 2
        elif kind == "local_ubyte":
            ins.ival = code[pc]; pc += 1
        elif kind == "iinc":
            ins.iinc_idx = code[pc]; pc += 1
            ins.iinc_const = _s1(code, pc); pc += 1
        elif kind == "branch2":
            off = _s2(code, pc); pc += 2
            ins.target = start + off
        elif kind == "branch4":
            off = _s4(code, pc); pc += 4
            ins.target = start + off
        elif kind == "atype":
            ins.atype = NEWARRAY_TYPES.get(code[pc], "int"); pc += 1
        elif kind == "invokeinterface":
            ins.cp_index = _u2(code, pc); pc += 2
            ins.count = code[pc]; pc += 1
            pc += 1
        elif kind == "invokedynamic":
            ins.cp_index = _u2(code, pc); pc += 2
            pc += 2
        elif kind == "multianewarray":
            ins.cp_index = _u2(code, pc); pc += 2
            ins.dims = code[pc]; pc += 1
        elif kind == "tableswitch":
            pad = (4 - (pc % 4)) % 4
            pc += pad
            default = _s4(code, pc); pc += 4
            low = _s4(code, pc); pc += 4
            high = _s4(code, pc); pc += 4
            targets = {}
            for val in range(low, high + 1):
                off = _s4(code, pc); pc += 4
                targets[val] = start + off
            targets[None] = start + default
            ins.targets = targets
        elif kind == "lookupswitch":
            pad = (4 - (pc % 4)) % 4
            pc += pad
            default = _s4(code, pc); pc += 4
            npairs = _s4(code, pc); pc += 4
            targets = {}
            for _ in range(npairs):
                match = _s4(code, pc); pc += 4
                off = _s4(code, pc); pc += 4
                targets[match] = start + off
            targets[None] = start + default
            ins.targets = targets
        elif kind == "wide":
            sub_op = code[pc]; pc += 1
            sub_entry = OPCODES.get(sub_op)
            sub_mn = sub_entry[0] if sub_entry else f"0x{sub_op:02x}"
            idx = _u2(code, pc); pc += 2
            if sub_op == 0x84:  # iinc
                const = _s2(code, pc); pc += 2
                ins.iinc_idx = idx
                ins.iinc_const = const
                ins.mnemonic = "iinc"
                ins.kind = "iinc"
            else:
                ins.ival = idx
                ins.mnemonic = sub_mn
                ins.kind = "local_ubyte"
        ins.next_pc = pc
        instrs[start] = ins
        order.append(start)
    return instrs, order

File: resources/engine_cpp/src/ir.cpp
// ir.cpp - см. ir.hpp. 1:1 порт ir.py.
#include "ir.hpp"

#include <cstdio>

namespace nd {

namespace {
int8_t s1_at(const std::vector<uint8_t>& b, size_t p) { return static_cast<int8_t>(b[p]); }
uint16_t u2_at(const std::vector<uint8_t>& b, size_t p) {
    return (uint16_t(b[p]) << 8) | uint16_t(b[p + 1]);
}
int16_t s2_at(const std::vector<uint8_t>& b, size_t p) { return static_cast<int16_t>(u2_at(b, p)); }
int32_t s4_at(const std::vector<uint8_t>& b, size_t p) {
    uint32_t v = (uint32_t(b[p]) << 24) | (uint32_t(b[p + 1]) << 16) | (uint32_t(b[p + 2]) << 8) | uint32_t(b[p + 3]);
    return static_cast<int32_t>(v);
}
}  // namespace

DecodedMethod decode_method(const std::vector<uint8_t>& code) {
    DecodedMethod out;
    size_t n = code.size();
    size_t pc = 0;

    while (pc < n) {
        size_t start = pc;
        uint8_t opcode = code[pc];
        const OpcodeInfo& entry = OPCODES[opcode];
        if (entry.mnemonic == nullptr) {
            Instruction ins;
            ins.pc = start;
            ins.opcode = opcode;
            char buf[32];
            std::snprintf(buf, sizeof(buf), "<unknown 0x%02x>", opcode);
            ins.mnemonic = buf;
            ins.kind = OperandKind::None;
            ins.next_pc = start + 1;
            out.instrs[start] = ins;
            out.order.push_back(start);
            pc += 1;
            continue;
        }
        std::string mnemonic = entry.mnemonic;
        OperandKind kind = entry.operand_kind;
        pc += 1;

        Instruction ins;
        ins.pc = start;
        ins.opcode = opcode;
        ins.mnemonic = mnemonic;
        ins.kind = kind;

        switch (kind) {
            case OperandKind::None:
                break;
            case OperandKind::Byte:
                ins.ival = s1_at(code, pc); pc += 1;
                break;
            case OperandKind::Short:
                ins.ival = s2_at(code, pc); pc += 2;
                break;
            case OperandKind::UByteCp:
                ins.cp_index = code[pc]; pc += 1;
                break;
            case OperandKind::UShortCp:
                ins.cp_index = u2_at(code, pc); pc += 2;
                break;
            case OperandKind::LocalUByte:
                ins.ival = code[pc]; pc += 1;
                break;
            case OperandKind::IInc:
                ins.iinc_idx = code[pc]; pc += 1;
                ins.iinc_const = s1_at(code, pc); pc += 1;
                break;
            case OperandKind::Branch2: {
                int16_t off = s2_at(code, pc); pc += 2;
                ins.target = static_cast<int64_t>(start) + off;
                break;
            }
            case OperandKind::Branch4: {
                int32_t off = s4_at(code, pc); pc += 4;
                ins.target = static_cast<int64_t>(start) + off;
                break;
            }
            case OperandKind::AType: {
                uint8_t t = code[pc]; pc += 1;
                auto it = NEWARRAY_TYPES.find(t);
                ins.atype = (it != NEWARRAY_TYPES.end()) ? it->second : std::string("int");
                break;
            }
            case OperandKind::InvokeInterface:
                ins.cp_index = u2_at(code, pc); pc += 2;
                ins.count = code[pc]; pc += 1;
                pc += 1;
                break;
            case OperandKind::InvokeDynamic:
                ins.cp_index = u2_at(code, pc); pc += 2;
                pc += 2;
                break;
            case OperandKind::MultiANewArray:
                ins.cp_index = u2_at(code, pc); pc += 2;
                ins.dims = code[pc]; pc += 1;
                break;
            case OperandKind::TableSwitch: {
                size_t pad = (4 - (pc % 4)) % 4;
                pc += pad;
                int32_t deflt = s4_at(code, pc); pc += 4;
                int32_t low = s4_at(code, pc); pc += 4;
                int32_t high = s4_at(code, pc); pc += 4;
                SwitchTargets targets;
                for (int64_t val = low; val <= static_cast<int64_t>(high); ++val) {
                    int32_t off = s4_at(code, pc); pc += 4;
                    targets.emplace_back(val, static_cast<int64_t>(start) + off);
                }
                targets.emplace_back(std::nullopt, static_cast<int64_t>(start) + deflt);
                ins.targets = std::move(targets);
                break;
            }
            case OperandKind::LookupSwitch: {
                size_t pad = (4 - (pc % 4)) % 4;
                pc += pad;
                int32_t deflt = s4_at(code, pc); pc += 4;
                int32_t npairs = s4_at(code, pc); pc += 4;
                SwitchTargets targets;
                for (int32_t i = 0; i < npairs; ++i) {
                    int32_t match = s4_at(code, pc); pc += 4;
                    int32_t off = s4_at(code, pc); pc += 4;
                    targets.emplace_back(match, static_cast<int64_t>(start) + off);
                }
                targets.emplace_back(std::nullopt, static_cast<int64_t>(start) + deflt);
                ins.targets = std::move(targets);
                break;
            }
            case OperandKind::Wide: {
                uint8_t sub_op = code[pc]; pc += 1;
                const OpcodeInfo& sub_entry = OPCODES[sub_op];
                std::string sub_mn;
                if (sub_entry.mnemonic != nullptr) {
                    sub_mn = sub_entry.mnemonic;
                } else {
                    char buf[8];
                    std::snprintf(buf, sizeof(buf), "0x%02x", sub_op);
                    sub_mn = buf;
                }
                uint16_t idx = u2_at(code, pc); pc += 2;
                if (sub_op == 0x84) {  // iinc
                    int16_t cst = s2_at(code, pc); pc += 2;
                    ins.iinc_idx = idx;
                    ins.iinc_const = cst;
                    ins.mnemonic = "iinc";
                    ins.kind = OperandKind::IInc;
                } else {
                    ins.ival = idx;
                    ins.mnemonic = sub_mn;
                    ins.kind = OperandKind::LocalUByte;
                }
                break;
            }
        }
        ins.next_pc = pc;
        out.instrs[start] = ins;
        out.order.push_back(start);
    }
    return out;
}

}  // namespace nd
